### DISCUSSION TITLE
Error in Basic Auth. Header Response in Middleware after upgrading to Next 12.2.5

### DISCUSSION THREAD

**Discussion by sdhacker** [vercel/next.js#39925] [Help] [answered]
Hi,
I recently upgraded Next from 12.1.6 to 12.2.5 after upgrading its giving "NestedMiddleware" error which i solved by moving _middlware.ts  to  root/middleware.ts but after moving it giving " A middleware can not alter response's body" error ,where i am returing basic realm authentication header as response but now when i am accessing particular path where it should ask authentication  its giving " Internal Server Error" 
can someone please help??

Code i am using in middleware.ts
```import type { NextFetchEvent, NextRequest } from 'next/server'
export function middleware(req: NextRequest, event: NextFetchEvent) {
  console.log('middleware')
  const pathName = req.nextUrl.pathname
  console.log('pathName:', pathName)
  if (pathName.startsWith('/api/push')) {
    const authorizationHeader = req.headers.get('authorization')
    console.log('authorizationHeader:', authorizationHeader)
    let wrongCredentials: boolean = false
    if (authorizationHeader) {
      const headerValueBase64 = authorizationHeader.split(' ')[1]
      const headerValue = Buffer.from(headerValueBase64, 'base64').toString()
      const [username, password] = headerValue.split(':')
      const basicAuthUsername = String(process.env.BASIC_AUTH_USERNAME)
      const basicAuthPassword = String(process.env.BASIC_AUTH_PASSWORD)
      if (username !== basicAuthUsername || password !== basicAuthPassword) {
        wrongCredentials = true
      }
    }
    console.log('wrongCredentials:', wrongCredentials)
    if (!authorizationHeader || wrongCredentials) {
      return new Response('401 Unauthorized', {
        status: 401,
        headers: {
          'WWW-Authenticate': 'Basic realm="Secure Area"',
        },
      })
    }
  }
}
```

**Comment by wwwdepot** [comment]
Next.js v12.2 made a few changes to Middleware API, one of them is

> Middleware can no longer produce a response body

What you can do is rewrite the request to an API route that produce the same HTTP 401 Response. Read the Middleware upgrade guide: https://nextjs.org/docs/messages/middleware-upgrade-guide

**Comment by icyJoseph** [comment, marked as answer]
As said above, middleware now honours the file routing system, so you need to redirect, next, or rewrite. 

I'd say, rewrite the request:

```ts
if (!authorizationHeader || wrongCredentials) {
   req.nextUrl.pathname = `/api/401-giving-api-route`;
   return Next.rewrite(req.nextUrl);
}
```

Then you implement `/api/401-giving-api-route`, to just respond with `401`, a bit of an annoying extra hop, but it should work. Haven't really tested it :), let me know how it goes.
